### PR TITLE
Replace removed GetEntityCountsForDebugging for Sync objects count P3A

### PR DESCRIPTION
**Pull request by AlexeyBarabash**
Resolves

This is a fix for https://github.com/brave/brave-core/pull/17450/commits/46c9d5982bda04cbd855d5aaed4d89d13e637bb7 

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/sync/service/brave_sync_service_impl.cc [hunk 1]
```diff
@@ -349,31 +350,67 @@ void BraveSyncServiceImpl::OnSyncCycleCompleted(
...
+          brave_sync::p3a::RecordSyncedObjectsCount(total_entities +
+                                                    known_to_sync_count.second);
+        },
+        total_entities));
+  } else {
+    brave_sync::p3a::RecordSyncedObjectsCount(total_entities);
+  }
+}
+
+void BraveSyncServiceImpl::GetEntityCountsForDebuggingBackport(
```
Comment (darkdh) [inline components/sync/service/brave_sync_service_impl.cc:380]: Looks like concern from upstream is legit, if one of the callback in barrier failed, we won't get entity count at all in that case. Can we keep track of receiving count along with its type and record the count once only when we receive from all in the repeated callback? When the recording happens, we also clear the tracking state.

Comment (darkdh) [reply components/sync/service/brave_sync_service_impl.cc:380]: so the main concern here is bound callback passed into `GetEntityCountsForDebuggingBackport` might not be called under some circumstances according to this [upstream change](https://chromium-review.googlesource.com/c/chromium/src/+/5388661)

File: components/sync/service/brave_sync_service_impl.cc [hunk 2]
```diff
@@ -349,31 +356,48 @@ void BraveSyncServiceImpl::OnSyncCycleCompleted(
...
+       GetUserSettings()->GetSelectedTypes()) {
+    ModelType model_type =
+        UserSelectableTypeToCanonicalModelType(user_selected_type);
+
+    ModelTypeController::TypeMap::const_iterator entry =
+        model_type_controllers_.find(model_type);
+    DCHECK(entry != model_type_controllers_.end())
+        << "Missing controller for type " << ModelTypeToDebugString(model_type);
+    if (entry != model_type_controllers_.end()) {
+      const std::unique_ptr<ModelTypeController>& controller = entry->second;
```
Comment (darkdh) [inline components/sync/service/brave_sync_service_impl.cc:371]: ModelTypeController* controller = entry->second.get();

**Review by darkdh — APPROVED** [review]
++

File: components/sync/service/brave_sync_service_impl.cc (hunk 2 above)
Comment (AlexeyBarabash) [reply components/sync/service/brave_sync_service_impl.cc]: thanks, fixed that at https://github.com/brave/brave-core/pull/22923/commits/9d0dfb159b6f3adbffda22364c125609dd298cf4

**Pull request merged**